pkg/sql/truncate.go

```diff
@@ -413,65 +415,93 @@ func reassignComment(
 		}
 	}
 
-	for i := range oldTableDesc.Columns {
-		id := oldTableDesc.Columns[i].ID
-		err = reassignColumnComment(ctx, p, oldTableDesc.ID, newTableDesc.ID, id)
-		if err != nil {
-			return err
-		}
+	if err := reassignAllColumnComments(ctx, p, oldTableDesc.ID, newTableDesc.ID, oldTableDesc.Columns); err != nil {
+		return err
 	}
 
-	for _, indexDesc := range oldTableDesc.Indexes {
-		err = reassignIndexComment(ctx, p, oldTableDesc.ID, newTableDesc.ID, indexDesc.ID)
-		if err != nil {
-			return err
-		}
+	if err := reassignAllIndexComments(ctx, p, oldTableDesc.ID, newTableDesc.ID, oldTableDesc.Indexes); err != nil {
+		return err
 	}
 
 	return nil
 }
 
-// reassignColumnComment reassign comment on column.
-func reassignColumnComment(
-	ctx context.Context, p *planner, oldID sqlbase.ID, newID sqlbase.ID, columnID sqlbase.ColumnID,
+// reassignAllColumnComments moves all column comments from oldID to newID.
+func reassignAllColumnComments(
+	ctx context.Context,
+	p *planner,
+	oldID sqlbase.ID,
+	newID sqlbase.ID,
+	columns []sqlbase.ColumnDescriptor,
 ) error {
-	comment, err := p.ExtendedEvalContext().ExecCfg.InternalExecutor.QueryRowEx(
+	if len(columns) == 0 {
+		return nil
+	}
+	var in strings.Builder
+	in.WriteString("(")
+	for i := range columns {
+		if i > 0 {
+			in.WriteString(", ")
+		}
+		in.WriteString(fmt.Sprintf("%d", columns[i].ID))
+	}
+	in.WriteString(")")
+	inString := in.String()
+
+	comments, err := p.ExtendedEvalContext().ExecCfg.InternalExecutor.QueryEx(
 		ctx,
-		"select-column-comment",
+		"select-column-comments",
 		p.txn,
 		sqlbase.InternalExecutorSessionDataOverride{User: security.RootUser},
-		`SELECT comment FROM system.comments WHERE type=$1 AND object_id=$2 AND sub_id=$3`,
+		fmt.Sprintf(
+			`SELECT sub_id, comment FROM system.comments WHERE type=$1 AND object_id=$2 AND sub_id IN %s`,
+			inString,
+		),
 		keys.ColumnCommentType,
 		oldID,
-		columnID)
+	)
 	if err != nil {
 		return err
 	}
+	columnToComment := make(map[sqlbase.ColumnID]tree.Datum)
+	for _, row := range comments {
+		colID := sqlbase.ColumnID(tree.MustBeDInt(row[0]))
+		columnToComment[colID] = row[1]
+	}
 
-	if comment != nil {
+	if len(columnToComment) > 0 {
+		var newVals strings.Builder
+		comma := ""
+		for colID, comment := range columnToComment {
+			newVals.WriteString(comma)
+			newVals.WriteString(fmt.Sprintf("(%d, %d, %d, %s)", keys.ColumnCommentType, newID, colID, comment))
```

Comment:
hmm this has my sql injection sensors running. does the library we use not handle multi-value VALUES?

Reply:
Yup, I'm worried about the same thing lol

Reply:
can you test it by having sql inject-y looking things?

Reply:
added -- it looks like the datum string formatter handles escaping of strings.